internal/provider/records_data_source.go

```diff
@@ -0,0 +1,176 @@
+							Computed:            true,
+						},
+						"id": schema.StringAttribute{
+							MarkdownDescription: "ID of this DNS record",
+							Computed:            true,
+						},
+						"name": schema.StringAttribute{
+							MarkdownDescription: "Name of this DNS record",
+							Computed:            true,
+						},
+						"ttl": schema.Int64Attribute{
+							MarkdownDescription: "Time to live of this record",
+							Computed:            true,
+						},
+						"type": schema.StringAttribute{
+							MarkdownDescription: "Type of this DNS record",
+							Computed:            true,
+						},
+						"value": schema.StringAttribute{
+							MarkdownDescription: "Value of this DNS record",
+							Computed:            true,
+						},
+					},
+				},
+			},
+			"zone_id": schema.StringAttribute{
+				MarkdownDescription: "ID of the DNS zone to get records from",
+				Required:            true,
+				Validators: []validator.String{
+					stringvalidator.LengthAtLeast(1),
+				},
+			},
+		},
+	}
+}
+
+func (d *recordsDataSource) Configure(_ context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	provider, ok := req.ProviderData.(*providerClient)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected *providerClient, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
+
+	d.provider = provider
+}
+
+func (d *recordsDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data recordsDataSourceModel
+
+	// Read Terraform configuration data into the model
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+
+	if data.ZoneID.ValueString() == "" {
+		resp.Diagnostics.AddError("Attribute Error", "no 'zone_id' set")
+	}
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	records, err := d.provider.apiClient.GetRecordsByZoneID(ctx, data.ZoneID.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("API Error", fmt.Sprintf("Unable to get records from zone, got error: %s", err))
+
+		return
+	}
+
+	// if zone == nil {
+	// 	resp.Diagnostics.AddError("API Error", fmt.Sprintf("DNS zone '%s' doesn't exist", data.Name.ValueString()))
+
+	// 	return
+	// }
+
+	// ns, diags := types.ListValueFrom(ctx, types.StringType, zone.NS)
+
+	// resp.Diagnostics.Append(diags...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var elements []recordDataSourceModel
+	for _, record := range *records {
+		elements = append(elements,
+			recordDataSourceModel{
+				ZoneID: types.StringValue(record.ZoneID),
+				ID:     types.StringValue(record.ID),
+				Type:   types.StringValue(record.Type),
+				Name:   types.StringValue(record.Name),
+				Value:  types.StringValue(record.Value),
+				TTL:    types.Int64Value(*record.TTL),  // FIXME memory error in this line
```

Comment:
```suggestion
				TTL:    types.Int64PointerValue(record.TTL),
```
TTL is an optional value, leading to a nil value which can raise a nil pointe panic.

Using `types.Int64PointerValue` could catch that behavior and return nil, if the pointer is nil